### DISCUSSION TITLE
Drawing optimizations

### DISCUSSION THREAD

**Discussion by mgnslndh** [AvaloniaUI/Avalonia#11129] [Q&A] [answered]
We're writing a control that will draw a lot of items using the [DrawGeometry](https://github.com/AvaloniaUI/Avalonia/blob/c21a87614204532d291adb3217f4d84cfda46097/src/Avalonia.Base/Media/DrawingContext.cs#L94) API. We do this by overriding the `void Render(DrawingContext context)` method and draw all the items every time the method is called. So far this is a naive proof of concept implementation and we feel that there are probably a lot things that can be improved to increase performance. 

Is it possible to use the `DrawingContext` to draw on something other then the screen, such as a bitmap perhaps, and cache that for later drawing on the screen's DrawingContext? This would enable us to not redraw all items each time the Render method is called, but instead just draw the bitmap, if the state has not changed. Is this even a good optimization? Anyone else done something similar?

**Comment by maxkatz6** [comment, marked as answer]
I see two possible alternative approaches.
1. Render to a bitmap, which then can be reused on each frame, as you mentioned. You can achieve this with RenderTargetBitmap which has draw APIs.
2. Go a lower level and implement ICustomDrawOperation, which will still render its content on each frame, but with slightly less overhead. You can create new immutable draw operation each time when your content is changed and reuse it on each frame. Possibly can access skia context directly.

First option, most likely, is more desirable in your case.
